File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/render/CodyAutoCompleteElementRenderer.java
package com.sourcegraph.cody.autocomplete.render;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorCustomElementRenderer;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.colors.EditorFontType;
import com.intellij.openapi.editor.impl.EditorImpl;
import com.intellij.openapi.editor.markup.TextAttributes;
import java.awt.*;
import java.util.Optional;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class CodyAutoCompleteElementRenderer implements EditorCustomElementRenderer {
  @NotNull public final String text;
  @NotNull protected final TextAttributes themeAttributes;
  @NotNull protected final Editor editor;
  @Nullable protected final AutoCompleteRendererType type;

  public CodyAutoCompleteElementRenderer(
      @NotNull String text, @NotNull Editor editor, @Nullable AutoCompleteRendererType type) {
    this.text = text;
    this.themeAttributes = AutoCompleteRenderUtils.getTextAttributesForEditor(editor);
    this.editor = editor;
    this.type = type;
  }

  @Override
  public int calcWidthInPixels(@NotNull Inlay inlay) {
    EditorImpl editor = (EditorImpl) inlay.getEditor();
    return editor.getFontMetrics(Font.PLAIN).stringWidth(text);
  }

  protected @NotNull Font getFont() {
    Font editorFont = this.editor.getColorsScheme().getFont(EditorFontType.PLAIN);
    return Optional.ofNullable(editorFont.deriveFont(Font.ITALIC)).orElse(editorFont);
  }

  protected int fontYOffset() {
    return (int) AutoCompleteRenderUtils.fontYOffset(getFont(), this.editor);
  }

  @Override
  public String toString() {
    return "CodyCompletionElementRenderer{" + "text='" + text + '\'' + '}';
  }

  public @NotNull String getText() {
    return this.text;
  }

  public @Nullable AutoCompleteRendererType getType() {
    return this.type;
  }
}

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/autocomplete/render/CodyAutocompleteElementRenderer.kt
package com.sourcegraph.cody.autocomplete.render

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.EditorCustomElementRenderer
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.editor.colors.EditorFontType
import com.intellij.openapi.editor.impl.EditorImpl
import com.intellij.openapi.editor.markup.TextAttributes
import com.sourcegraph.cody.vscode.InlineAutocompleteItem
import com.sourcegraph.config.ConfigUtil.getCustomAutocompleteColor
import com.sourcegraph.config.ConfigUtil.isCustomAutocompleteColorEnabled
import java.awt.Font
import java.util.function.Supplier

abstract class CodyAutocompleteElementRenderer(
    val text: String,
    val completionItems: List<InlineAutocompleteItem>,
    protected val editor: Editor,
    val type: AutocompleteRendererType
) : EditorCustomElementRenderer {
  protected val themeAttributes: TextAttributes

  init {
    val textAttributesFallback = Supplier {
      AutocompleteRenderUtil.getTextAttributesForEditor(editor)
    }
    themeAttributes =
        if (isCustomAutocompleteColorEnabled())
            getCustomAutocompleteColor()?.let {
              AutocompleteRenderUtil.getCustomTextAttributes(editor, it)
            }
                ?: textAttributesFallback.get()
        else textAttributesFallback.get()
  }

  override fun calcWidthInPixels(inlay: Inlay<*>): Int {
    val editor = inlay.editor as EditorImpl
    return editor.getFontMetrics(Font.PLAIN).stringWidth(text)
  }

  protected val font: Font
    get() {
      val editorFont = editor.colorsScheme.getFont(EditorFontType.PLAIN)
      return editorFont.deriveFont(Font.ITALIC) ?: editorFont
    }

  protected fun fontYOffset(): Int = AutocompleteRenderUtil.fontYOffset(font, editor).toInt()
}
